ledger-tool/src/main.rs

```diff
@@ -2973,3 +3022,65 @@ fn main() {
     measure_total_execution_time.stop();
     info!("{}", measure_total_execution_time);
 }
+
+fn record_transactions(
+    recv: crossbeam_channel::Receiver<TransactionStatusMessage>,
+    slots: Arc<Mutex<Vec<BankHashSlotDetails>>>,
+) {
+    for tsm in recv {
+        if let TransactionStatusMessage::Batch(batch) = tsm {
+            let slot = batch.bank.slot();
+
+            assert_eq!(batch.transactions.len(), batch.execution_results.len());
+
+            let transactions: Vec<_> = batch
+                .transactions
+                .iter()
+                .enumerate()
+                .map(|(no, tx)| {
+                    let message = tx.message();
+
+                    let accounts: Vec<String> = message
+                        .account_keys()
+                        .iter()
+                        .map(|acc| acc.to_string())
+                        .collect();
+
+                    let instructions = message
+                        .instructions()
+                        .iter()
+                        .map(|ix| UiInstruction::parse(ix, &message.account_keys(), None))
+                        .collect();
+
+                    let execution_results = batch.execution_results[no].clone();
+
+                    let is_simple_vote_tx = tx.is_simple_vote_transaction();
+
+                    BankHashTransaction {
+                        accounts,
+                        instructions,
+                        is_simple_vote_tx,
+                        execution_results,
+                        index: batch.transaction_indexes[no],
+                    }
+                })
+                .collect();
+
+            let mut slots = slots.lock().unwrap();
+
+            if let Some(recorded_slot) = slots.iter_mut().find(|f| f.slot == slot) {
+                recorded_slot.transactions.extend(transactions);
+            } else {
+                slots.push(BankHashSlotDetails {
+                    slot,
+                    transactions,
+                    ..Default::default()
+                });
+            }
```

Comment:
To confirm - this logic is grouping transactions by slot as long as the crossbeam channel is connected. Once the sender goes out of scope, we'll break out of `for tsm in recv` and get here. This means all of the serialization will occur once we've finished?

Reply:
Yes, this exactly right.

This does increase memory usage, and serialization on the fly might be better to prevent OOM. However, this means serializing the data bit by bit, and I haven't found a great way of doing this in rust. All I've found so far is that we can print the opening `[` and closing `]` manually and then adding `SlotDetails` per slot. That works but also needs `,` handling - no `,` for the last entry. It all seems a lot of busy work so I don't know if it's worthwhile or if there is a better method.